File: crates/core/src/light.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LightColor {
    Sunlight,
    Red,
    Green,
    Blue,
}

impl LightColor {
    pub fn from(color: usize) -> LightColor {
        match color {
            0 => LightColor::Sunlight,
            1 => LightColor::Red,
            2 => LightColor::Green,
            3 => LightColor::Blue,
            _ => panic!("Invalid light color!"),
        }
    }
}

pub struct LightUtils;

impl LightUtils {
    #[inline]
    pub fn extract_sunlight(light: u32) -> u32 {
        (light >> 12) & 0xF
    }

    #[inline]
    pub fn insert_sunlight(light: u32, level: u32) -> u32 {
        (light & 0x0FFF) | ((level & 0xF) << 12)
    }

    #[inline]
    pub fn extract_red_light(light: u32) -> u32 {
        (light >> 8) & 0xF
    }

    #[inline]
    pub fn insert_red_light(light: u32, level: u32) -> u32 {
        (light & 0xF0FF) | ((level & 0xF) << 8)
    }

    #[inline]
    pub fn extract_green_light(light: u32) -> u32 {
        (light >> 4) & 0xF
    }

    #[inline]
    pub fn insert_green_light(light: u32, level: u32) -> u32 {
        (light & 0xFF0F) | ((level & 0xF) << 4)
    }

    #[inline]
    pub fn extract_blue_light(light: u32) -> u32 {
        light & 0xF
    }

    #[inline]
    pub fn insert_blue_light(light: u32, level: u32) -> u32 {
        (light & 0xFFF0) | (level & 0xF)
    }

    #[inline]
    pub fn extract_all(light: u32) -> (u32, u32, u32, u32) {
        (
            Self::extract_sunlight(light),
            Self::extract_red_light(light),
            Self::extract_green_light(light),
            Self::extract_blue_light(light),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sunlight_roundtrip() {
        for level in 0..=15 {
            let light = LightUtils::insert_sunlight(0, level);
            assert_eq!(LightUtils::extract_sunlight(light), level);
        }
    }

    #[test]
    fn test_red_light_roundtrip() {
        for level in 0..=15 {
            let light = LightUtils::insert_red_light(0, level);
            assert_eq!(LightUtils::extract_red_light(light), level);
        }
    }

    #[test]
    fn test_green_light_roundtrip() {
        for level in 0..=15 {
            let light = LightUtils::insert_green_light(0, level);
            assert_eq!(LightUtils::extract_green_light(light), level);
        }
    }

    #[test]
    fn test_blue_light_roundtrip() {
        for level in 0..=15 {
            let light = LightUtils::insert_blue_light(0, level);
            assert_eq!(LightUtils::extract_blue_light(light), level);
        }
    }

    #[test]
    fn test_combined_lights() {
        let mut light = 0u32;
        light = LightUtils::insert_sunlight(light, 15);
        light = LightUtils::insert_red_light(light, 10);
        light = LightUtils::insert_green_light(light, 5);
        light = LightUtils::insert_blue_light(light, 3);

        assert_eq!(LightUtils::extract_sunlight(light), 15);
        assert_eq!(LightUtils::extract_red_light(light), 10);
        assert_eq!(LightUtils::extract_green_light(light), 5);
        assert_eq!(LightUtils::extract_blue_light(light), 3);
    }

    #[test]
    fn test_extract_all() {
        let mut light = 0u32;
        light = LightUtils::insert_sunlight(light, 12);
        light = LightUtils::insert_red_light(light, 8);
        light = LightUtils::insert_green_light(light, 4);
        light = LightUtils::insert_blue_light(light, 2);

        let (sun, red, green, blue) = LightUtils::extract_all(light);
        assert_eq!(sun, 12);
        assert_eq!(red, 8);
        assert_eq!(green, 4);
        assert_eq!(blue, 2);
    }

    #[test]
    fn test_light_color_from() {
        assert_eq!(LightColor::from(0), LightColor::Sunlight);
        assert_eq!(LightColor::from(1), LightColor::Red);
        assert_eq!(LightColor::from(2), LightColor::Green);
        assert_eq!(LightColor::from(3), LightColor::Blue);
    }
}

File: packages/ts-core/src/light.ts
import { toUint32 } from "./constants";

export enum LightColor {
  Sunlight = "sunlight",
  Red = "red",
  Green = "green",
  Blue = "blue",
}

type NumericLikeValue = number | string | boolean | object | null | undefined;

const toFiniteNumberOrZero = (value: NumericLikeValue): number => {
  if (typeof value === "number") {
    return Number.isFinite(value) ? value : 0;
  }

  try {
    const numericValue = Number(value);
    return Number.isFinite(numericValue) ? numericValue : 0;
  } catch {
    return 0;
  }
};

const toUint32WordOrZero = (value: NumericLikeValue): number => {
  return toFiniteNumberOrZero(value) >>> 0;
};

export const lightColorFromIndex = (color: number): LightColor => {
  switch (color) {
    case 0:
      return LightColor.Sunlight;
    case 1:
      return LightColor.Red;
    case 2:
      return LightColor.Green;
    case 3:
      return LightColor.Blue;
    default:
      throw new RangeError("Invalid light color!");
  }
};

export class LightUtils {
  static extractSunlight(light: number): number {
    return (toUint32WordOrZero(light) >>> 12) & 0xf;
  }

  static insertSunlight(light: number, level: number): number {
    const lightWord = toUint32WordOrZero(light);
    const normalizedLevel = toFiniteNumberOrZero(level);
    return toUint32((lightWord & 0x0fff) | ((normalizedLevel & 0xf) << 12));
  }

  static extractRedLight(light: number): number {
    return (toUint32WordOrZero(light) >>> 8) & 0xf;
  }

  static insertRedLight(light: number, level: number): number {
    const lightWord = toUint32WordOrZero(light);
    const normalizedLevel = toFiniteNumberOrZero(level);
    return toUint32((lightWord & 0xf0ff) | ((normalizedLevel & 0xf) << 8));
  }

  static extractGreenLight(light: number): number {
    return (toUint32WordOrZero(light) >>> 4) & 0xf;
  }

  static insertGreenLight(light: number, level: number): number {
    const lightWord = toUint32WordOrZero(light);
    const normalizedLevel = toFiniteNumberOrZero(level);
    return toUint32((lightWord & 0xff0f) | ((normalizedLevel & 0xf) << 4));
  }

  static extractBlueLight(light: number): number {
    return toUint32WordOrZero(light) & 0xf;
  }

  static insertBlueLight(light: number, level: number): number {
    const lightWord = toUint32WordOrZero(light);
    const normalizedLevel = toFiniteNumberOrZero(level);
    return toUint32((lightWord & 0xfff0) | (normalizedLevel & 0xf));
  }

  static extractAll(light: number): [number, number, number, number] {
    return [
      LightUtils.extractSunlight(light),
      LightUtils.extractRedLight(light),
      LightUtils.extractGreenLight(light),
      LightUtils.extractBlueLight(light),
    ];
  }
}

export interface LightChannels {
  sunlight: number;
  red: number;
  green: number;
  blue: number;
}

type LightChannelRecord = {
  sunlight?: number;
  red?: number;
  green?: number;
  blue?: number;
};

const toLightChannelRecordOrNull = (
  channels: Partial<LightChannels> | null | undefined
): LightChannelRecord | null => {
  return channels !== null && typeof channels === "object"
    ? (channels as LightChannelRecord)
    : null;
};

const safeReadLightChannel = (
  channels: LightChannelRecord | null,
  key: keyof LightChannelRecord
): number | undefined => {
  if (channels === null) {
    return undefined;
  }

  try {
    return channels[key];
  } catch {
    return undefined;
  }
};

export class Light {
  static sunlight(light: number): number {
    return LightUtils.extractSunlight(light);
  }

  static red(light: number): number {
    return LightUtils.extractRedLight(light);
  }

  static green(light: number): number {
    return LightUtils.extractGreenLight(light);
  }

  static blue(light: number): number {
    return LightUtils.extractBlueLight(light);
  }

  static withSunlight(light: number, level: number): number {
    return LightUtils.insertSunlight(light, level);
  }

  static withRed(light: number, level: number): number {
    return LightUtils.insertRedLight(light, level);
  }

  static withGreen(light: number, level: number): number {
    return LightUtils.insertGreenLight(light, level);
  }

  static withBlue(light: number, level: number): number {
    return LightUtils.insertBlueLight(light, level);
  }

  static pack(channels: Partial<LightChannels>): number {
    let light = 0;
    const normalizedChannels = toLightChannelRecordOrNull(channels);
    const sunlight = safeReadLightChannel(normalizedChannels, "sunlight");
    const red = safeReadLightChannel(normalizedChannels, "red");
    const green = safeReadLightChannel(normalizedChannels, "green");
    const blue = safeReadLightChannel(normalizedChannels, "blue");

    if (sunlight !== undefined) {
      light = Light.withSunlight(light, sunlight);
    }

    if (red !== undefined) {
      light = Light.withRed(light, red);
    }

    if (green !== undefined) {
      light = Light.withGreen(light, green);
    }

    if (blue !== undefined) {
      light = Light.withBlue(light, blue);
    }

    return light;
  }

  static unpack(light: number): LightChannels {
    return {
      sunlight: Light.sunlight(light),
      red: Light.red(light),
      green: Light.green(light),
      blue: Light.blue(light),
    };
  }
}
